### PR TITLE
ROX-16279: Add new flag to diagnostic bundle

### PR DESCRIPTION
**Pull request by fredrb**
## Description

Allows diagnostic bundles to have alert information.

**Context:** the reason why this is being added is to check whether Sensors with re-sync disabled are behaving as expected, without having to re-run sensor with re-sync enabled and compare alerts. The alternative we are proposing here is: when requesting a diagnostic bundle, allow a flag that will request alerts, reprocess all deployments and request alerts again. What we (as team Maple) can do with this info is: check that there is no delta between the two. This gives us confidence that, even after reprocessing everything, Sensor was able to identify all resource changes that led to alerts. 

This flag should be used only through `roxctl` and it's here as a sanity check to test that secured clusters with re-sync disabled are producing the correct alerts. 

Since the new files (`alerts-*.json`) will take a while to generate, we shouldn't have them enabled by default, and should only be used explicitly by customers that are testing the re-sync changes. 

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- [ ] Request diagnostic bundle via `roxctl`

### CODE REVIEW

File: roxctl/central/debug/download_diagnostics.go [hunk 1]
```diff
@@ -69,6 +73,7 @@ To specify timeout, run  'roxctl' command:
 	c.PersistentFlags().StringVar(&outputDir, "output-dir", "", "output directory in which to store bundle")
 	c.PersistentFlags().StringSliceVar(&clusters, "clusters", nil, "comma separated list of sensor clusters from which logs should be collected")
 	c.PersistentFlags().StringVar(&since, "since", "", "timestamp starting when logs should be collected from sensor clusters")
+	c.PersistentFlags().BoolVar(&withAlerts, "with-alerts", false, "add alert data before/after reassessing policies")
```
Comment (0x656b694d) [inline roxctl/central/debug/download_diagnostics.go:76]: Not clear to me what does "before/after" mean. Before and/or after?

File: central/debug/service/service.go [hunk 2]
```diff
@@ -598,6 +620,30 @@ func (s *serviceImpl) writeZippedDebugDump(ctx context.Context, w http.ResponseW
...
+					ReprocessDeployments: &central.ReprocessDeployments{},
+				},
+			})
+			if err != nil {
+				log.Warnf("Failed to request reprocess deployment for cluster: %s", conn.ClusterID())
+				continue
+			}
+		}
+
+		log.Info("Requested reprocess clusters. Waiting for a minute before finish writing the diagnostic bundle")
```
Comment (ivan-degtiarenko) [inline central/debug/service/service.go:639]: Why 1 minute? Is there a guarantee it will be enough?

File: central/debug/service/service.go [hunk 3]
```diff
@@ -598,6 +620,30 @@ func (s *serviceImpl) writeZippedDebugDump(ctx context.Context, w http.ResponseW
 
 	fetchAndAddJSONToZip(debugDumpCtx, zipWriter, "system-configuration.json", s.getConfig)
 
+	// Fetch alerts and append to the zip
+	if opts.withAlerts {
+		fetchAndAddJSONToZip(debugDumpCtx, zipWriter, "alerts-before.json", s.fetchAlerts)
+		// clusterSet := set.NewFrozenStringSet(opts.clusters...)
+		for _, conn := range s.sensorConnMgr.GetActiveConnections() {
+			err := conn.InjectMessage(debugDumpCtx, &central.MsgToSensor{
+				Msg: &central.MsgToSensor_ReprocessDeployments{
```
Comment (ivan-degtiarenko) [inline central/debug/service/service.go:629]: I wonder if we are OK with introducing side effects to the diag bundle, and whether we shouldn't rather introduce completely separate flow for that

File: central/debug/service/service.go [hunk 4]
```diff
@@ -598,6 +620,30 @@ func (s *serviceImpl) writeZippedDebugDump(ctx context.Context, w http.ResponseW
...
+				log.Warnf("Failed to request reprocess deployment for cluster: %s", conn.ClusterID())
+				continue
+			}
+		}
+
+		log.Info("Requested reprocess clusters. Waiting for a minute before finish writing the diagnostic bundle")
+
+		// We have to sleep here to make sure that any new alerts that could be generated after requesting all deployments
+		// to be reprocessed will appear in the new JSON file.
+		time.Sleep(time.Minute)
```
Comment (ivan-degtiarenko) [inline central/debug/service/service.go:643]: During this minute context deadline will be ignored

**Review by ivan-degtiarenko — CHANGES_REQUESTED** [review]
I noticed that the ticket mentions UI changes as well. Is it planned to be done in a separate PR?

File: roxctl/central/debug/download_diagnostics.go (hunk 1 above)
Comment (fredrb) [reply roxctl/central/debug/download_diagnostics.go:76]: I think "before and after" will be more clear. These are the files that will be output in the diagnostic bundle. Alerts before and after reassessing policies.

File: central/debug/service/service.go (hunk 2 above)
Comment (fredrb) [reply central/debug/service/service.go:639]: This is pretty arbitrary. The current approach was to set to a time where we're are confident that most alerts would have arrived after reassessing. But we don't have any way to guarantee that all _possible_ alerts were already received. 

We thought also about making this duration configurable. E.g. with a flag `--reprocess-delay 30s`?

File: central/debug/service/service.go (hunk 3 above)
Comment (fredrb) [reply central/debug/service/service.go:629]: I agree. But adding a new flow might be inconvenient for customers. So, our proposal here is to add a flag that is disabled by default in diag bundles and will only be used for this specific case. 
```
roxctl central debug download-diagnostics --alerts=true --timeout=100
```

Ideally, we should remove this after this release, once re-sync is disabled by default on customer's clusters.  

Would you suggest adding a new `roxctl debug` command altogether for this?

**Comment by fredrb** [comment]
@ivan-degtiarenko @0x656b694d sorry folks, you were automatically added as reviewers without much context on what we were doing. I've updated the description of the PR with the reasoning on why we are adding this. Hope it clarifies. If you have more questions or suggestions let me know. 



> I noticed that the ticket mentions UI changes as well. Is it planned to be done in a separate PR?

We aren't doing any UI changes anymore. That was the original plan, but now we are going `roxctl` only.

**Comment by fredrb** [comment]
Closing in favor of #5559

**Comment by fredrb** [comment]
Closing in favor of #5559

**Pull request closed without merging**